### PR TITLE
build: Do not include server symbols in wallet

### PR DESCRIPTION
**Pull request by maflcko**
This reverts a hacky workaround from commit b83cc0f, which only happens to work due to compiler optimizations. Then, it actually fixes the linker error.

The underlying problem is that the wallet includes symbols from the server (ui_interface), which usually results in linker failures. Though, in this specific case the linker failures have not been observed (unless `-O0`) because our compilers were smart enough to strip unused symbols.

Fix the underlying problem by creating a new header-only with the needed symbol and move ui_interface to node to clarify that this is part of libbitcoin_server.

### CODE REVIEW

**Review by hebasto — COMMENTED** [review]
Approach ACK, tested with GCC 7.5.0 -- works as expected.

> The underlying problem is that the wallet includes symbols from the server (ui_interface), which _usually_ results in linker failures.

Trying to get the complete understanding of linker failures. Could you elaborate or point out further reading the reasons of linker failures.

From my current understanding, if a symbol is exported, it could be linked to, no?

File: src/util/ui_change_type.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2012-2020 The Bitcoin Core developers
```
Comment (hebasto) [inline src/util/ui_change_type.h:1]: nit:
```suggestion
// Copyright (c) 2020 The Bitcoin Core developers
```

File: src/qt/bitcoin.cpp [hunk 2]
```diff
@@ -29,12 +29,12 @@
 
 #include <interfaces/handler.h>
 #include <interfaces/node.h>
+#include <node/ui_interface.h>
```
Comment (hebasto) [inline src/qt/bitcoin.cpp:32]: It seems `qt/bitcoin.cpp` does not use any symbols from `node/ui_interface.h` directly.

File: src/qt/splashscreen.cpp [hunk 3]
```diff
@@ -12,9 +12,9 @@
 #include <interfaces/handler.h>
 #include <interfaces/node.h>
 #include <interfaces/wallet.h>
+#include <node/ui_interface.h>
```
Comment (hebasto) [inline src/qt/splashscreen.cpp:15]: It seems `qt/splashscreen.cpp` does not use any symbols from `node/ui_interface.h` directly.

File: src/util/ui_change_type.h (hunk 1 above)
Comment (MarcoFalke) [reply src/util/ui_change_type.h:1]: The enum has been added in 2012

**Comment by MarcoFalke** [comment]
> Trying to get the complete understanding of linker failures. Could you elaborate or point out further reading the reasons of linker failures.

*  bitcoin-wallet tool: Drop libbitcoin_server.a dependency #15639 
*  Move-only: Pull wallet code out of libbitcoin_server #15638

File: src/qt/bitcoin.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/qt/bitcoin.cpp:32]: thx, removed

File: src/qt/splashscreen.cpp (hunk 3 above)
Comment (MarcoFalke) [reply src/qt/splashscreen.cpp:15]: thx, removed

**Comment by MarcoFalke** [comment]
Removed unused includes as requested by @hebasto

**Review by hebasto — APPROVED** [review]
ACK fa30d36c5a94cd2f71185162b577e05ab9adf717, tested on Linux Mint 19.3 (x86_64) with both GCC 7.5.0 and Clang 6.0.0, using `-O0` compiler option.

**Comment by practicalswift** [comment]
Concept ACK

Thanks for fixing! This `-O0` gotcha has annoyed me (mildly) for a while :)

**Comment by laanwj** [comment]
This looks reasonable. It's important to be consistent about which libraries are allowed to use which symbols. Concept ACK.

**Comment by MarcoFalke** [comment]
Rebased

**Comment by Sjors** [comment]
ACK faca730

**Comment by MarcoFalke** [comment]
@hebasto Mind to re-ACK?

**Review by hebasto — APPROVED** [review]
re-ACK faca73000fa8975c28f6be8be01957c1ae94ea62, since the [previous](https://github.com/bitcoin/bitcoin/pull/19331#pullrequestreview-434420539) review:
- rebased
- added "ci: Install fixed version of clang-format for linters" commit (related to #19095)

**Comment by laanwj** [comment]
ACK faca73000fa8975c28f6be8be01957c1ae94ea62

**Pull request merged**